src/main/webapp/script.js

```diff
@@ -69,37 +79,132 @@ function toggleAQLayer() {
 }
 
 function populateAQVisualisationData() {
-    fetch("/visualisation").then(response => response.json()).then((data) => {
-        const aqData = [];
-        for (i = 0; i < data.length; i++) {
-            const dataItem = data[i];
-            let dataPoint = {};
-            dataPoint.location = new google.maps.LatLng(data[i].lat,data[i].long);
-            dataPoint.weight = dataItem.aqi;
-            aqData.push(dataPoint);
-        }
-        aqData.push({location: new google.maps.LatLng(-37.785, 150.435), weight: 3})
+    let fetchURL = "/visualisation";
+    fetchURL += `?zoom-level=${map.getZoom()}`;
+    const mapBounds = map.getBounds();
+    const swCorner = mapBounds.getSouthWest();
+    const neCorner = mapBounds.getNorthEast();
+    fetchURL += `&sw-lat=${swCorner.lat()}`;
+    fetchURL += `&sw-long=${swCorner.lng()}`;
+    fetchURL += `&ne-lat=${neCorner.lat()}`;
+    fetchURL += `&ne-long=${neCorner.lng()}`;
+    fetch(fetchURL).then(response => response.json()).then((data) => {
```

Comment:
Gather the values, then compose the fetchURL?

Reply:
Is that in case the zoom gets changed mid-way through creating it?

Reply:
String formatting operations a fairly quick; gathering all the calls together doesn't reduce the window for the context to change by a whole lot (microseconds either way).

However:
```
let fetchURL = "/visualisation?zoom-level=${map.getZoom()}&sw-lat=${swCorner.lat()}&sw-long=${swCorner.lng()}&ne-lat=${neCorner.lat()}&ne-long=${neCorner.lng()}";
```
Is tidier, and keeps the template string all in one line.

Also, in various languages, string manipulation is more expensive when broken up. I'm not sure on the Javascript efficiency of strings, but in Go a string is static, and manipulation of the string causes the allocation of a new static string. Memory allocation is very expensive, so doing this as few times as possible is a big efficiency win.

Reply:
I've changed it to one line